app/models/measurement_unit.rb

```diff
@@ -29,15 +29,27 @@ def units(unit_code, unit_key)
     end
 
     def weight_units
-      measurement_units.select { |_k, v| v['measurement_unit_type'] == 'weight' }.keys
+      @weight_units ||= begin
+        units = measurement_units.select { |_k, v| v['measurement_unit_type'] == 'weight' }.keys
+
+        Set.new(units)
+      end
     end
 
     def volume_units
-      measurement_units.select { |_k, v| v['measurement_unit_type'] == 'volume' }.keys
+      @volume_units ||= begin
+        units = measurement_units.select { |_k, v| v['measurement_unit_type'] == 'volume' }.keys
+
+        Set.new(units)
+      end
     end
 
     def percentage_abv_units
-      measurement_units.select { |_k, v| v['measurement_unit_type'] == 'percentage_abv' }.keys
+      @percentage_abv_units ||= begin
+        units = measurement_units.select { |_k, v| v['measurement_unit_type'] == 'percentage_abv' }.keys
+
+        Set.new(units)
+      end
     end
```

Comment:
I see a similar pattern here, 
maybe we can simplify those `*_units` methods with something like this:

```
    def get_units_for(unit_type)
      units = measurement_units.select { |_k, v| v['measurement_unit_type'] == unit_type }.keys
      Set.new(units)
    end
```
and then call it 
`get_units_for('weight')` 
for example.

Of course, this is valid if the memoisation is not strictly necessary.

Reply:
We definitely want the class level memoisation IMO since these will get called super frequently

Reply:
I think explicit is better for a pattern only used in one file, though